### PR TITLE
feature(zjow): envpoool env example in new pipeline

### PR DESCRIPTION
**Pull request by zjowowen**

### CODE REVIEW

File: ding/envs/env_manager/envpool_env_manager.py [hunk 1]
```diff
@@ -55,18 +72,29 @@ def launch(self) -> None:
             seed = 0
         else:
             seed = self._seed
+
+        kwargs = {}
+        if "episodic_life" in self._cfg:
```
Comment (PaParaZz1) [inline ding/envs/env_manager/envpool_env_manager.py:77]: add default value in default config, so we can omit the if expression and directly conduct the assignment operation

File: ding/envs/env_manager/envpool_env_manager.py [hunk 2]
```diff
@@ -39,10 +54,12 @@ def default_config(cls) -> EasyDict:
         # Async mode: batch_size <  env_num
         env_num=8,
         batch_size=8,
+        image_observation=True,
     )
 
     def __init__(self, cfg: EasyDict) -> None:
-        self._cfg = cfg
+        self._cfg = self.default_config()
+        self._cfg.update(cfg)
```
Comment (PaParaZz1) [inline ding/envs/env_manager/envpool_env_manager.py:62]: why merge config here, we have already merged the config of env manager in `compile_config` function

File: ding/envs/env_manager/envpool_env_manager.py [hunk 3]
```diff
@@ -124,3 +156,158 @@ def env_num(self) -> int:
...
+            return self._action_space
+        except AttributeError:
+            self.launch()
+            self.close()
+            return self._action_space
+
+
+@ENV_MANAGER_REGISTRY.register('env_pool_v4')
+class PoolEnvManagerV2:
+    '''
```
Comment (PaParaZz1) [inline ding/envs/env_manager/envpool_env_manager.py:181]: """

File: ding/envs/env_manager/envpool_env_manager.py [hunk 4]
```diff
@@ -124,3 +156,158 @@ def env_num(self) -> int:
...
+    def action_space(self) -> 'gym.spaces.Space':  # noqa
+        try:
+            return self._action_space
+        except AttributeError:
+            self.launch()
+            self.close()
+            return self._action_space
+
+
+@ENV_MANAGER_REGISTRY.register('env_pool_v4')
```
Comment (PaParaZz1) [inline ding/envs/env_manager/envpool_env_manager.py:179]: why v4 here

File: ding/envs/env_manager/envpool_env_manager.py [hunk 5]
```diff
@@ -124,3 +156,158 @@ def env_num(self) -> int:
...
+        - Atari: "Pong-v5", "SpaceInvaders-v5", "Qbert-v5"
+        - Classic Control: "CartPole-v0", "CartPole-v1", "Pendulum-v1"
+    '''
+
+    @classmethod
+    def default_config(cls) -> EasyDict:
+        return EasyDict(deepcopy(cls.config))
+
+    config = dict(
+        type='envpool',
```
Comment (PaParaZz1) [inline ding/envs/env_manager/envpool_env_manager.py:197]: modify type and make it as same as what it used in registry

File: ding/envs/env_manager/envpool_env_manager.py [hunk 6]
```diff
@@ -124,3 +156,158 @@ def env_num(self) -> int:
...
+    def reset(self) -> None:
+        self._envs.async_reset()
+        ready_obs = {}
+        while True:
+            obs, _, _, info = self._envs.recv()
+            env_id = info['env_id']
+            obs = obs.astype(np.float32)
+            if self._cfg.image_observation:
+                obs /= 255.0
+            ready_obs = deep_merge_dicts({i: o for i, o in zip(env_id, obs)}, ready_obs)
```
Comment (PaParaZz1) [inline ding/envs/env_manager/envpool_env_manager.py:257]: directly use the assignment operation, don't use `deep_merge_dicts` too many times, it is a complex function

File: ding/envs/env_manager/envpool_env_manager.py [hunk 7]
```diff
@@ -124,3 +156,158 @@ def env_num(self) -> int:
...
+
+    @property
+    def action_space(self) -> 'gym.spaces.Space':  # noqa
+        try:
+            return self._action_space
+        except AttributeError:
+            self.launch()
+            self.close()
+            self._ready_obs = {}
+            return self._action_space
```
Comment (PaParaZz1) [inline ding/envs/env_manager/envpool_env_manager.py:313]: add envpooltest for this file

File: ding/example/dqn_envpool_nstep.py [hunk 8]
```diff
@@ -0,0 +1,139 @@
+import datetime
```
Comment (PaParaZz1) [inline ding/example/dqn_envpool_nstep.py:1]: rename to dqn_nstep_envpool.py

File: ding/example/dqn_envpool_nstep.py [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+import torch
+try:
+    torch.multiprocessing.set_start_method('spawn')
```
Comment (PaParaZz1) [inline ding/example/dqn_envpool_nstep.py:4]: why add this

File: ding/example/dqn_envpool_nstep.py [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+if __name__ == "__main__":
+
+    import argparse
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--seed", type=int, default=0, help="random seed")
+    parser.add_argument("--collector_env_num", type=int, default=8, help="collector env number")
+    parser.add_argument("--collector_batch_size", type=int, default=8, help="collector batch size")
+    arg = parser.parse_args()
+
+    pong_dqn_envpool_config.env.collector_env_num = arg.collector_env_num
```
Comment (PaParaZz1) [inline ding/example/dqn_envpool_nstep.py:125]: use an new config for envpool demo

File: ding/policy/common_utils.py [hunk 11]
```diff
@@ -1,9 +1,11 @@
 from typing import List, Any, Dict, Callable
+import numpy as np
 import torch
 import numpy as np
 import treetensor.torch as ttorch
 from ding.utils.data import default_collate
-from ding.torch_utils import to_tensor, to_ndarray, unsqueeze, squeeze
+from ding.torch_utils import to_tensor, to_ndarray, unsqueeze, squeeze, to_device
+import time
```
Comment (PaParaZz1) [inline ding/policy/common_utils.py:8]: polish import sequence

File: ding/worker/collector/sample_serial_collector.py [hunk 12]
```diff
@@ -34,7 +34,8 @@ def __init__(
             policy: namedtuple = None,
             tb_logger: 'SummaryWriter' = None,  # noqa
             exp_name: Optional[str] = 'default_experiment',
-            instance_name: Optional[str] = 'collector'
+            instance_name: Optional[str] = 'collector',
+            timer_cuda: bool = False,
```
Comment (PaParaZz1) [inline ding/worker/collector/sample_serial_collector.py:38]: add corresponding comments about the newly added argument

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: ding/envs/env_manager/envpool_env_manager.py (hunk 2 above)
Comment (zjowowen) [reply ding/envs/env_manager/envpool_env_manager.py:62]: There are two env manager config, one for evaluator and one for collector. It's too complicated to use compile_config with auto=True.
I suggest use  compile_config with auto=False.

File: ding/envs/env_manager/envpool_env_manager.py (hunk 1 above)
Comment (zjowowen) [reply ding/envs/env_manager/envpool_env_manager.py:77]: fixed.

File: ding/envs/env_manager/envpool_env_manager.py (hunk 3 above)
Comment (zjowowen) [reply ding/envs/env_manager/envpool_env_manager.py:181]: fixed

File: ding/envs/env_manager/envpool_env_manager.py (hunk 4 above)
Comment (zjowowen) [reply ding/envs/env_manager/envpool_env_manager.py:179]: fixed

File: ding/example/dqn_envpool_nstep.py (hunk 8 above)
Comment (zjowowen) [reply ding/example/dqn_envpool_nstep.py:1]: fixed

File: ding/envs/env_manager/envpool_env_manager.py (hunk 5 above)
Comment (zjowowen) [reply ding/envs/env_manager/envpool_env_manager.py:197]: fixed, thanks

File: ding/example/dqn_envpool_nstep.py (hunk 9 above)
Comment (zjowowen) [reply ding/example/dqn_envpool_nstep.py:4]: fixed

File: ding/policy/common_utils.py (hunk 11 above)
Comment (zjowowen) [reply ding/policy/common_utils.py:8]: fixed

File: ding/worker/collector/sample_serial_collector.py (hunk 12 above)
Comment (zjowowen) [reply ding/worker/collector/sample_serial_collector.py:38]: fixed

File: ding/example/dqn_envpool_nstep.py (hunk 10 above)
Comment (zjowowen) [reply ding/example/dqn_envpool_nstep.py]: fixed

File: ding/envs/env_manager/envpool_env_manager.py (hunk 6 above)
Comment (zjowowen) [reply ding/envs/env_manager/envpool_env_manager.py:248]: fixed

File: ding/envs/env_manager/envpool_env_manager.py (hunk 7 above)
Comment (zjowowen) [reply ding/envs/env_manager/envpool_env_manager.py:304]: fixed

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

**Comment by PaParaZz1** [comment]
This PR will be continually updated in #753.

**Pull request closed without merging**